packages/network-controller/src/NetworkController.ts

```diff
@@ -896,6 +986,234 @@ export class NetworkController extends BaseControllerV2<
+        const autoManagedNetworkClient = createAutoManagedNetworkClient(
+          networkClientConfiguration,
+        );
+        registry[AutoManagedNetworkClientCategory.BuiltIn][id] =
+          autoManagedNetworkClient;
+      } else {
+        if (networkClientConfiguration.chainId === undefined) {
+          throw new Error('chainId must be provided for custom RPC endpoints');
+        }
+        if (networkClientConfiguration.rpcUrl === undefined) {
+          throw new Error('rpcUrl must be provided for custom RPC endpoints');
+        }
+        const autoManagedNetworkClient = createAutoManagedNetworkClient(
+          networkClientConfiguration,
+        );
+        registry[AutoManagedNetworkClientCategory.Custom][id] =
+          autoManagedNetworkClient;
+      }
+    }
+
+    return registry;
+  }
+
+  /**
+   * Constructs the list of network clients for built-in networks (that is,
+   * those that we know Infura supports).
+   *
+   * @returns The network clients.
+   */
+  #buildIdentifiedBuiltInNetworkClientConfigurations(): [
+    string,
+    InfuraNetworkClientConfiguration,
+  ][] {
+    return knownKeysOf(InfuraNetworkType).map((network) => {
+      const networkClientConfiguration: InfuraNetworkClientConfiguration = {
+        type: NetworkClientType.Infura,
+        network,
+        infuraProjectId: this.#infuraProjectId,
+      };
+      return [network, networkClientConfiguration];
+    });
+  }
+
+  /**
+   * Constructs the list of network clients for custom networks (that is, those
+   * that have been added to `networkConfigurations` in state).
+   *
+   * @returns The network clients.
+   */
+  #buildIdentifiedCustomNetworkClientConfigurations(): [
+    string,
+    CustomNetworkClientConfiguration,
+  ][] {
+    return Object.entries(this.state.networkConfigurations).map(
+      ([id, networkConfiguration]) => {
+        const networkClientConfiguration: CustomNetworkClientConfiguration = {
+          type: NetworkClientType.Custom,
+          chainId: networkConfiguration.chainId,
+          rpcUrl: networkConfiguration.rpcUrl,
+        };
+        return [id, networkClientConfiguration];
+      },
+    );
+  }
+
+  /**
+   * Converts the provider config object in state to a network client
+   * configuration object.
+   *
+   * @returns The network client config.
+   * @throws If the provider config is of type "rpc" and lacks either a
+   * `chainId` or an `rpcUrl`.
+   */
+  #buildIdentifiedNetworkClientConfigurationFromProviderConfig(): [
+    string,
+    NetworkClientConfiguration,
+  ] {
+    const { providerConfig } = this.state;
+
+    if (providerConfig.type === NetworkType.rpc) {
+      if (providerConfig.chainId === undefined) {
+        throw new Error('chainId must be provided for custom RPC endpoints');
+      }
+      if (providerConfig.rpcUrl === undefined) {
+        throw new Error('rpcUrl must be provided for custom RPC endpoints');
+      }
+      const id =
```

Comment:
Should we try to find a matching custom network configuration? e.g. one with the same RPC URL. Today the RPC URL is effectively a unique identifier.

Otherwise we might accidentally duplicate it, creating two clients for the same thing. Then when the user switches away and back to this network, we'd lose the caches in the middleware layer.

Reply:
Hmm... I wonder if what we might really want is to change the `providerConfig` in state so that it's equal to its matching network configuration (if one exists)? The `rollbackToPreviousProvider` method uses `providerConfig` to know what to go back to, so if we don't create a network client for `providerConfig` but we also don't change `providerConfig`, then `rollbackToPreviousProvider` will break completely.

Reply:
Yes, that could work. I commented on this line because I was thinking that instead of assigning an id when it's missing, we'd generate an ID that matches a pre-existing network configuration/client. Basically the same as your idea I think, except without any state changes.

Reply:
That's a good suggestion. I've made that change here, and added tests to confirm the intended behavior: [24aa829](https://github.com/MetaMask/core/pull/1439/commits/24aa8295a11d062b7cb24ee933f867bd9955163c)